packages/block/src/block.ts

```diff
@@ -299,11 +316,76 @@ export class Block {
     }
   }
 
-  private _validateUncleHeader(uncleHeader: BlockHeader, blockchain: Blockchain) {
-    // TODO: Validate that the uncle header hasn't been included in the blockchain yet.
-    // This is not possible in ethereumjs-blockchain since this PR was merged:
-    // https://github.com/ethereumjs/ethereumjs-blockchain/pull/47
-    const height = this.header.number
-    return uncleHeader.validate(blockchain, height)
+  private async _validateUncleHeaders(uncleHeaders: BlockHeader[], blockchain: Blockchain) {
+    if (uncleHeaders.length == 0) {
+      return
+    }
+
+    // validate the header
+    for (let i = 0; i < uncleHeaders.length; i++) {
+      const header = uncleHeaders[i]
+      await header.validate(blockchain, this.header.number)
+    }
+
+    const canonicalBlockMap: Block[] = []
+
+    let lowestUncleNumber = this.header.number.clone()
+
+    uncleHeaders.map((header) => {
+      if (header.number.lt(lowestUncleNumber)) {
+        lowestUncleNumber = header.number.clone()
+      }
+    })
+
+    const getBlocks = this.header.number.clone().sub(lowestUncleNumber).addn(1).toNumber()
+
+    // See Geth: https://github.com/ethereum/go-ethereum/blob/b63bffe8202d46ea10ac8c4f441c582642193ac8/consensus/ethash/consensus.go#L207
+    let parentHash = this.header.parentHash
+    for (let i = 0; i < getBlocks; i++) {
+      const parentBlock = await this._getBlockByHash(blockchain, parentHash)
+      if (!parentBlock) {
+        throw new Error('could not find parent block')
+      }
+      canonicalBlockMap.push(parentBlock)
+      parentHash = parentBlock.header.parentHash
+    }
+
+    canonicalBlockMap.map((block) => {
```

Comment:
It would be good if these different validation sections in `_validateUncleHeaders` get some short comment headers (maybe analogue to the already commented `validateUncles` starting section) to structure this a bit and grasp what is going on.

Reply:
Yes good point, will do this.

Reply:
Have squashed this logic using some helpers, should be easier to read now 😄